goaci.go

```diff
@@ -38,45 +41,75 @@ func debug(i ...interface{}) {
 	}
 }
 
+type StringVector []string
+
+func (v *StringVector) String() string {
+	return "\"" + strings.Join(*v, "\" \"") + "\""
+}
+
+func (v *StringVector) Set(str string) error {
+	*v = append(*v, str)
+	return nil
+}
+
 func main() {
+	var (
+		execOpts            StringVector
+		goDefaultBinaryDesc string
+		goBinaryOpt         string
+		goPathOpt           string
+	)
+
+	// Find the go binary
+	gocmd, err := exec.LookPath("go")
+
+	flag.Var(&execOpts, "exec", "Parameters passed to app, can be used multiple times")
+	if err != nil {
+		goDefaultBinaryDesc = "Go binary to use (default: none found in $PATH, so it must be provided)"
+	} else {
+		goDefaultBinaryDesc = "Go binary to use (default: whatever go in $PATH)"
+	}
+	flag.StringVar(&goBinaryOpt, "go-binary", gocmd, goDefaultBinaryDesc)
+	flag.StringVar(&goPathOpt, "go-path", "", "Custom GOPATH (default: a temporary directory)")
+	flag.Parse()
 	if os.Getenv("GOPATH") != "" {
-		die("to avoid confusion GOPATH must not be set")
+		warn("GOPATH envvar is ignored, use --go-path=\"$GOPATH\" option instead")
 	}
-	goroot := os.Getenv("GOROOT")
-	if goroot == "" {
-		die("GOROOT must be set")
+	if os.Getenv("GOROOT") != "" {
+		warn("GOROOT envvar is ignored, use --go-binary=\"$GOROOT/bin/go\" option instead")
 	}
 	if os.Getenv("GOACI_DEBUG") != "" {
 		Debug = true
 	}
 
+	if goBinaryOpt == "" {
+		die("go binary not found")
+	}
+
 	// Set up a temporary directory for everything (gopath and builds)
 	tmpdir, err := ioutil.TempDir("", "goaci")
 	if err != nil {
 		die("error setting up temporary directory: %v", err)
 	}
 	defer os.RemoveAll(tmpdir)
+	if goPathOpt == "" {
+		goPathOpt = tmpdir
+	}
 
 	// Scratch build dir for aci
 	acidir := filepath.Join(tmpdir, "aci")
 
-	// Be explicit with gobin
+	// Let's put final binary in tmpdir
 	gobin := filepath.Join(tmpdir, "bin")
 
-	// Find the go binary
-	gocmd, err := exec.LookPath("go")
-	if err != nil {
-		die("could not find `go` in path")
-	}
-
 	// Construct args for a go get that does a static build
-	// TODO(jonboulle): go version 1.4
 	args := []string{
-		gocmd,
+		goBinaryOpt,
 		"get",
 		"-a",
 		"-tags", "netgo",
 		"-ldflags", "'-w'",
+		"-installsuffix", "netgo",
```

Comment:
Heh, no idea. I think it does not hurt (for narrow definition of hurt, installsuffix has its own problems) - currently it didn't work for me, because my 1.3 is provided by distro, so I'm getting some permission problems (please see the main description of this PR).

I'll build `go 1.3` from source tomorrow and test it.